enhancements/baremetal/ztp-metal3.md

```diff
@@ -0,0 +1,325 @@
+  and the baremetal-operator's hardware inventory format (which, in turn, is
+  derived from the Ironic's inventory format).
+
+- The *standard* flow is highly [extensible][hardware managers] via the
+  [ironic-agent-image][ironic-agent] container image. This extensibility does
+  not apply to the *live ISO* flow since it does not use the Ironic agent.
+
+- Any different path through the `BareMetalHost` state machine makes it
+  exponentially more difficult for upstream developers to reason about the
+  effects of any changes. In addition, the *live iso* flow is not tested
+  upstream by the Metal3 project.
+
+### Goals
+
+- ZTP follows the *standard* Metal3 flow instead of *live ISO*.
+- (i)PXE boot, firwmare and RAID settings work for ZTP without introducing
+  special API in Metal3 components.
+- Continue to allow non-ZTP assisted deployments without depending on Metal3.
+- Not depend on assisted components in the core metal platform
+
+### Non-Goals
+
+- Converging two agent implementations (Ironic and assisted) into a single one.
+- Supporting per-host network configuration through the BMH
+  PreprovisioningNetworkData API in the ZTP case.
+- Performing firmware config changes only after the assisted service runs host
+  validations.
+
+## Proposal
+
+- Building CoreOS images for the *standard* flow is handled by
+  [image-customization-controller][icc] (ICC). ICC will start ignoring all
+  `PreprovisioningImage`s that have the `infraenvs.agent-install.openshift.io`
+  label (copied from the corresponding `BareMetalHost` by BMO).
+
+- A new controller will be created as part of the assisted service to
+  handle `PreprovisioningImage`s that do have the InfraEnv label. Its exact
+  implementation will be defined by the assisted team. The assisted agent
+  configuration will be updated to not start the agent automatically in the
+  presence of the Ironic agent.
+
+- To boot CoreOS via network, three artifacts are required: the kernel, the
+  initramfs and the root filesystem. In the current ICC implementation, only
+  the initramfs is a dynamically generated artifact, while the kernel and the
+  rootfs are taken from the payload on the Metal3 pod start-up and used for
+  all nodes. Since ZTP supports installing clusters of a different version and
+  even architecture from the hub cluster, we will add separate fields
+  to the `PreprovisioningImage`'s `Status`:
+
+  `KernelURL` to store the matching kernel.
+
+  `ExtraKernelParams` to provide a link to the root file system. BMO will be
+  updated to pass the parameters to the node.
+
+- [Ironic agent][ironic-agent] currently contains a custom deployment plugin
+  (*custom deploy method* in baremetal-operator terms, *custom deploy step* in
+  Ironic terms) that invokes `coreos-installer` for the image installation.
+  A new custom deploy method `start_assisted_install` will be added to start
+  the assisted agent at the right moment.
```

Comment:
yeah, a bit of hand-waving here. I don't know BMAC well enough to say for sure when it starts the deployment.

Reply:
I guess this will happen the same way BMAC  updates the BMH Image.URL today (by looking the InfraEnv label on the BMH)
[https://github.com/openshift/assisted-service/blob/4eee641dccec201432e154189a67b09[…]19186e8/internal/controller/controllers/bmh_agent_controller.go](https://github.com/openshift/assisted-service/blob/4eee641dccec201432e154189a67b09f719186e8/internal/controller/controllers/bmh_agent_controller.go#L678)
It should work just fine.

Reply:
From my understanding, BMAC adds the custom deploy method as soon as it first reconciles a BMH that is linked to an InfraEnv.  This has the effect of IPA starting Assisted Agent "at the right moment", which means after any bios/raid/whatever configurations have been completed - correct?